### PR TITLE
[QA-185] build openssl from source for macos

### PR DESCRIPTION
**Pull request by naushniki**
This PR is intended to get rid of installed OpenSSL dependency for macOS nodes in CI: therefore oskar will download and compile necessary OpenSSL on it's own unless there is explicit (non-CI) need to still use a local one.

QA-185 task in JIRA tracks this.

### CODE REVIEW

**Comment by naushniki** [comment]
https://jenkins.arangodb.biz:3456/view/Mac/job/arangodb-ANY-mac-packages/2107/

**Comment by naushniki** [comment]
https://jenkins.arangodb.biz:3456/view/Mac/job/arangodb-ANY-mac-packages/2108/

**Comment by naushniki** [comment]
https://jenkins.arangodb.biz:3456/view/Mac/job/arangodb-ANY-mac-packages/2109/

**Comment by naushniki** [comment]
https://jenkins.arangodb.biz:3456/view/Mac/job/arangodb-ANY-mac-matrix/2588/

**Comment by KVS85** [comment]
PR:
- [ ] devel: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/18837/
- [ ] 3.9: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/18838/
- [ ] 3.8: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/18839/
- [ ] 3.7: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/18840/

Packages:
- [ ] devel: https://jenkins.arangodb.biz/view/Mac/job/arangodb-ANY-mac-packages/2186/
- [ ] 3.9: https://jenkins.arangodb.biz/view/Mac/job/arangodb-ANY-mac-packages/2187/
- [ ] 3.8: https://jenkins.arangodb.biz/view/Mac/job/arangodb-ANY-mac-packages/2188/
- [ ] 3.7: https://jenkins.arangodb.biz/view/Mac/job/arangodb-ANY-mac-packages/2189/

**Review by KVS85 — APPROVED** [review]
LGTM.

File: helper.mac.fish [hunk 1]
```diff
@@ -77,22 +81,98 @@ function findRequiredMinMacOS
...
-  set -l oversion $argv[1]
+  set -gx OPENSSL_VERSION $argv[1]
+end
 
-  if test "$oversion" = ""
-    set -e OPENSSL_VERSION
-    return 0
+function downloadOpenSSL
+  if test -z $OPENSSL_VERSION
+    echo "Need OPENSSL_VERSION global variable!"
```
Comment (mchacki) [inline helper.mac.fish:89]: Please improve this helper output.
What should be the content of the Variable? What is the "default"?

File: helper.mac.fish [hunk 2]
```diff
@@ -77,22 +81,98 @@ function findRequiredMinMacOS
...
+  curl -LO $url
+  rm -rf openssl-$OPENSSL_VERSION
+  tar -xzvf openssl-$OPENSSL_VERSION.tar.gz
+end
 
-    case '1.1.1'
-      set -gx OPENSSL_VERSION $oversion
+function buildOpenSSL
+  if test "$OPENSSL_SOURCE_DIR" = ""
+    echo "OPENSSL_SOURCE_DIR is not defined. Please download source codes before building OpenSSL!"
```
Comment (mchacki) [inline helper.mac.fish:105]: How do I download?
it seems "run fish> downloadOpenSSL" should be printed out here to improve the message

File: helper.mac.fish [hunk 3]
```diff
@@ -77,22 +81,98 @@ function findRequiredMinMacOS
...
+    return 1
+  else if ! test -d $OPENSSL_SOURCE_DIR
+      echo "Directory $OPENSSL_SOURCE_DIR does not exist!"
```
Comment (mchacki) [inline helper.mac.fish:108]: Why not just create it?

File: helper.mac.fish [hunk 4]
```diff
@@ -77,22 +81,98 @@ function findRequiredMinMacOS
...
+      return 1
+  end
+  cd $OPENSSL_SOURCE_DIR
+  mkdir build
+  
+  if test -z "$ARCH"
+    echo "ARCH is not set! Can't decide wether to build OpenSSL for arm64 or x86_64."
```
Comment (mchacki) [inline helper.mac.fish:115]: Why cannot not detect my System arch? And just build it?

File: helper.mac.fish [hunk 5]
```diff
@@ -77,22 +81,98 @@ function findRequiredMinMacOS
...
+  set -xg OPENSSL_ROOT $OPENSSL_SOURCE_DIR/build  
+
+  switch $BUILDMODE
+    case "Debug"
+      set mode debug
+    case "Release" "RelWithDebInfo" "MinSizeRel"
+      set mode release
     case '*'
-      echo "unknown openssl version $oversion"
+      echo "Unknown BUILDMODE value: $BUILDMODE. Can't choose OpenSSL build type."
```
Comment (mchacki) [inline helper.mac.fish:150]: Please print out available build modes here. And print out how to properly set it

**Review by mchacki — CHANGES_REQUESTED** [review]
Please update the README file.

There are some new input variables added that are not described in the README.
It is unclear for a user what values they should have, what their effects are.
Also there now is some way to build openSSL from Scratch, and some way to include a precompiled version. This needs to be written into the README.

From a usability perspective can we try the following:
If the variable is not set, can we check if we a have an open ssl in the standard home-brew location, if we find it we take it, if we not find out we ask for the variable to be provided?

As a developer I would prefer this script to work without setting anything from my side on a "standard" setup.

Eg. DEPLYOMENT_TARGET should default to what I have on my system. (out of scope for this PR!)
OPEN_SSL should use the default brew location.
etc.

THe technical changes look reasonable, but I could not test it as I did not know what effects to expect.

**Comment by mchacki** [comment]
On a side note:
Please fill in some information in the description of the PR.
What is this PR supposed to do?
Is there a bug fixed by it?
Is there a new feature added?

This helps all reviewers that are not involved directly with this PR to understand what should be achieved and to judge if it is achieved by this PR.
Thank you :+1:

File: helper.mac.fish (hunk 1 above)
Comment (dothebart) [reply helper.mac.fish:89]: shouldn't this use the contents of the `VERSIONS` file in the source: 
```
OPENSSL_MACOS   "1.1.1l"
```

File: helper.mac.fish (hunk 2 above)
Comment (KVS85) [reply helper.mac.fish]: Fixed.

File: helper.mac.fish (hunk 1 above)
Comment (KVS85) [reply helper.mac.fish]: Fixed.

File: helper.mac.fish (hunk 4 above)
Comment (KVS85) [reply helper.mac.fish:109]: It's detected right at the beginning. It's just a matter of checking whether it's still good (the environment wasn't changed or we initialized oskar at some weird place).

File: helper.mac.fish (hunk 3 above)
Comment (KVS85) [reply helper.mac.fish]: Refactored to not deal with it.

File: helper.mac.fish (hunk 5 above)
Comment (KVS85) [reply helper.mac.fish]: Fixed.

**Comment by KVS85** [comment]
> On a side note: Please fill in some information in the description of the PR. What is this PR supposed to do? Is there a bug fixed by it? Is there a new feature added?
> 
> This helps all reviewers that are not involved directly with this PR to understand what should be achieved and to judge if it is achieved by this PR. Thank you 👍

Fixed: https://github.com/arangodb/oskar/pull/360#issue-787252557.

**Comment by KVS85** [comment]
> Please update the README file.
> 
> There are some new input variables added that are not described in the README. It is unclear for a user what values they should have, what their effects are. Also there now is some way to build openSSL from Scratch, and some way to include a precompiled version. This needs to be written into the README.
> 
> From a usability perspective can we try the following: If the variable is not set, can we check if we a have an open ssl in the standard home-brew location, if we find it we take it, if we not find out we ask for the variable to be provided?
> 
> As a developer I would prefer this script to work without setting anything from my side on a "standard" setup.
> 
> Eg. DEPLYOMENT_TARGET should default to what I have on my system. (out of scope for this PR!) OPEN_SSL should use the default brew location. etc.
> 
> THe technical changes look reasonable, but I could not test it as I did not know what effects to expect.

Added Homebrew OpenSSL detection.

**Comment by KVS85** [comment]
PR:
- [ ] devel: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/18861/
- [ ] 3.9: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/18862/
- [ ] 3.8: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/18863/
- [ ] 3.7: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/18864/

Packages:
- [ ] devel: https://jenkins.arangodb.biz/view/Mac/job/arangodb-ANY-mac-packages/2190/
- [ ] 3.9: https://jenkins.arangodb.biz/view/Mac/job/arangodb-ANY-mac-packages/2191/
- [ ] 3.8: https://jenkins.arangodb.biz/view/Mac/job/arangodb-ANY-mac-packages/2192/
- [ ] 3.7: https://jenkins.arangodb.biz/view/Mac/job/arangodb-ANY-mac-packages/2193/

**Comment by KVS85** [comment]
PR:
- [ ] devel: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/18871/
- [ ] 3.9: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/18872/
- [ ] 3.8: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/18873/
- [ ] 3.7: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/18874/

Packages:
- [ ] devel: https://jenkins.arangodb.biz/view/Mac/job/arangodb-ANY-mac-packages/2190/
- [ ] 3.9: https://jenkins.arangodb.biz/view/Mac/job/arangodb-ANY-mac-packages/2191/
- [ ] 3.8: https://jenkins.arangodb.biz/view/Mac/job/arangodb-ANY-mac-packages/2192/
- [ ] 3.7: https://jenkins.arangodb.biz/view/Mac/job/arangodb-ANY-mac-packages/2193/

**Comment by KVS85** [comment]
PR:
- [ ] devel: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/18875/
- [ ] 3.9: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/18876/
- [ ] 3.8: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/18877/

**Review by mchacki — APPROVED** [review]
Thanks a lot.
Tested it locally as well, with installed OpenSSL and self compiled OpenSSL.

Seems to work on my machine :+1:

Also thanks for improving all the error messages.

**Review by mchacki — CHANGES_REQUESTED** [review]
Sorry need to reject again, retried and the latest version does NOT work anymore.

oskarOpenSSL
buildStaticArangoDB -DUSE_FAILURE_TESTS=On
Results in:
<download and unzip of OpenSSL (works)
Couldn't find OpenSSL 1.1.1l at /Users/mchacki/Library/Projects/oskar/work/openssl/openssl-1.1.1l!
Please download OpenSSL source with `downloadOpenSSL` function before building it!
Build error!

**Comment by KVS85** [comment]
> Sorry need to reject again, retried and the latest version does NOT work anymore.
> 
> oskarOpenSSL buildStaticArangoDB -DUSE_FAILURE_TESTS=On Results in: <download and unzip of OpenSSL (works) Couldn't find OpenSSL 1.1.1l at /Users/mchacki/Library/Projects/oskar/work/openssl/openssl-1.1.1l! Please download OpenSSL source with `downloadOpenSSL` function before building it! Build error!

@mchacki Fixed.

**Comment by KVS85** [comment]
PR:
- [ ] devel: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/18910/
- [ ] 3.9: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/18911/
- [ ] 3.8: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/18912/
- [ ] 3.7: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/18913/

Packages:
- [ ] devel: https://jenkins.arangodb.biz/view/Mac/job/arangodb-ANY-mac-packages/2201/
- [ ] 3.9: https://jenkins.arangodb.biz/view/Mac/job/arangodb-ANY-mac-packages/2202/
- [ ]  3.8: https://jenkins.arangodb.biz/view/Mac/job/arangodb-ANY-mac-packages/2203/
- [ ]  3.7: https://jenkins.arangodb.biz/view/Mac/job/arangodb-ANY-mac-packages/2204/

**Review by mchacki — APPROVED** [review]
Works for me again.

On a side note, it seems we then build OpenSSL with -O0 (so disabled optimization) is this intentional?

**Comment by KVS85** [comment]
> Works for me again.
> 
> On a side note, it seems we then build OpenSSL with -O0 (so disabled optimization) is this intentional?

We build all modes: debug and release, shared and static. `-O0` is for the debug part.

**Pull request merged**